### PR TITLE
Fixing config file lock mechanism

### PR DESCRIPTION
**Pull request by sverdlov93**
- [ ] All [tests](https://github.com/jfrog/jfrog-cli-core#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [ ] This pull request is on the dev branch.
- [ ] I used gofmt for formatting the code before submitting the pull request.
-----
Fix config commands locking mechanism.
Make all config commands that write to config file  - **atomic**. 

Also, get lock's file modified time from os, instead of using time.Now() (as suggested by @HWoidt)

should fix https://github.com/jfrog/jfrog-cli-core/issues/350

### CODE REVIEW

**Comment by sverdlov93** [comment]
Hi @HWoidt,
First of all, Thanks for reporting this issue and for your very informative logs and data.
Our apologies for the inconvenience caused by this.

I am currently working on the fix, and have some more changes and tests to add.
I will update here as soon as it's done and ready for merge.

**Comment by sverdlov93** [comment]
Hi, @HWoidt.
The PR is waiting for review but if you want to test it on your machine you can pull [JFrog CLI PR]( https://github.com/sverdlov93/jfrog-cli/tree/config-concurrency-test) and run `./build/build.sh` to get the executable.

File: common/commands/config.go
Comment (eyalbe4) [inline common/commands/config.go:24]: CmdType --> ConfigAction

File: common/commands/config.go
Comment (eyalbe4) [inline common/commands/config.go:45]: cmdType --> configAction

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

**Review by HWoidt — COMMENTED** [review]
Hi @sverdlov93 ,
Sorry for letting you wait. I had other stuff on my plate.
I have now tested your changes on our Ubuntu 20.04 machine with the bash script I provided earlier:
1. When running tasks concurrently the command often fails because the Stat-call (lock.go:169) on an already deleted lock-file causes the command to fail instead of being retried. I changed this locally and made the following observations with the Stat error ignored.
2. There sometimes is a deadlock, both processes are waiting until their lock-retry loop runs out after ~2 minutes and then return with a failure. I suspect this to be due to two files having the same timestamp. This tie can occur with higher frequency than before because the filesystem timestamps have [lower resolution](https://stackoverflow.com/questions/14392975/timestamp-accuracy-on-ext4-sub-millsecond).
3. The corruption can still be observed, although with vastly lower probability. This occurred maybe 2 times in the 3h that I left it running with 40 concurrent processes. I fear that we are running into the same conceptual problem as with the previous solution: The file-creation and timestamp generation are not within the same critical section. Probably the filesystem first prepares all file meta-data (including the timestamp) and only then commits the changes to disk, allowing for a time window in which another process can run through the whole sequence of creating the file and checking all timestamps. Since this occurs within the same system call, the time between timestamp-creation and visibility of the new file is shorter, resulting in a lower probability of hitting this issue.

With your changes I see much reduced probability for this to occur. We initially ran into this after about 1 Month, while starting three concurrent jf-config processes at the same time once a day. Considering that in the tests the error now happened after more than 30min compared to the 10-20s it was before and with 4 times as many threads, we can assume at least a factor of 100x reduction in probability. On our CI server we have circumvented this issue now by assigning individual config files per worker, using the JFROG_CLI_HOME_DIR environment variable.

For correctness it would probably be best to use a mechanism explicitly designed for file locking (i.e. LockFileEx on windows and flock on linux). There is already a cross platform go implementation used for exactly the same usecase of protecting configfiles in the "go" commandline tool: [go.internal.lockedfile](https://pkg.go.dev/cmd/go/internal/lockedfile). Unfortunately this is an internal package. There is a project containing useful internal packages from the go standard library which [exports](https://github.com/rogpeppe/go-internal/tree/master/lockedfile) this though: https://github.com/rogpeppe/go-internal

File: common/commands/config.go
Comment (eyalbe4) [inline common/commands/config.go:109]: Not supported... --> Unsupported...

File: common/commands/config.go
Comment (eyalbe4) [inline common/commands/config.go:237]: ///  --> //

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**
